### PR TITLE
Robot Field Localisation

### PR DESCRIPTION
**Pull request by JakeFountain**
This update implements a particle filter for tracking the robot's location relative to the field. In particular we track an offset from the odometry position maintained by the robot

### CODE REVIEW

File: module/localisation/RobotParticleLocalisation/src/RobotModel.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        arma::vec::fixed<size> limitState(const arma::vec::fixed<size>& state);
+
+        arma::mat::fixed<size, size> processNoise();
+
+        arma::vec3 processNoiseDiagonal;
+
+        // number and range of reset particles
+        int n_rogues          = 0;
+        arma::vec3 resetRange = {10, 10, 6};
```
Comment (josiahw) [inline module/localisation/RobotParticleLocalisation/src/RobotModel.h]: Be careful here - allowing the full range for heading for rogues will likely end up with the robot flipping direction sometimes. Can this be prevented by assuming a rogue is +/- only a small heading from the current estimate?

File: shared/utility/math/filter/ParticleFilter.h [hunk 2]
```diff
@@ -60,20 +60,40 @@ namespace math {
...
-            void setState(StateVec initialMean, StateMat initialCovariance) {
+            void resetAmbiguous(std::vector<StateVec> initialMean,
+                                std::vector<StateMat> initialCovariance,
+                                int number_of_particles_ = 100) {
+                if (initialMean.size() != initialCovariance.size()) {
+                    throw std::runtime_error(std::string(__FILE__) + " : " + std::to_string(__LINE__)
+                                             + " different number of means vs covariances provided.");
+                }
+                // Sample the same number of particles for each possibility
+                particles =
```
Comment (josiahw) [inline shared/utility/math/filter/ParticleFilter.h]: This and other consecutive joins would be much faster and cleaner as a lambda (if this doesn't compile Trent will know how to make it work):
particles = arma::mat(number_of_particles_ / initialMean.size(), initialMean.size());
int i = 0;
particles.each_col([&i](arma::col& result) {
result =  getParticles(initialMean[i], initialCovariance[i], number_of_particles_ / initialMean.size());
++i;
});

File: module/localisation/RobotParticleLocalisation/src/RobotModel.h (hunk 1 above)
Comment (JakeFountain) [inline module/localisation/RobotParticleLocalisation/src/RobotModel.h]: This is the default range. The robot is configured to have zero rogues, so it wont affect it. I think I will default configure (0,0,0) for the reset range too though

File: module/support/NUbugger/src/Localisation.cpp [hunk 3]
```diff
@@ -39,105 +39,19 @@ namespace support {
     using message::localisation::Model;
 
     void NUbugger::provideLocalisation() {
+
         handles["localisation"].push_back(
-            on<Every<100, std::chrono::milliseconds>,
-               Optional<With<std::vector<Ball>>>,
-               Optional<With<std::vector<Self>>>,
+            on<Every<10, Per<std::chrono::seconds>>,
```
Comment (TrentHouliston) [inline module/support/NUbugger/src/Localisation.cpp]: Change this to trigger functions. `on<Trigger<Self>>` and `on<Trigger<Ball>>`.
Having them grouped like this is pointless and won't update as fast as it could.

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 4]
```diff
@@ -115,12 +120,18 @@ namespace vision {
                 MEASUREMENT_LIMITS_RIGHT = config["measurement_limits"]["right"].as<uint>();
                 MEASUREMENT_LIMITS_TOP   = config["measurement_limits"]["top"].as<uint>();
                 MEASUREMENT_LIMITS_BASE  = config["measurement_limits"]["base"].as<uint>();
+
+                vector3_covariance = config["vector3_covariance"].as<arma::vec>();
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: Match the style of the other configuration parameters, either change all the others or these two :P

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 5]
```diff
@@ -239,6 +250,7 @@ namespace vision {
                         }
                     }
 
+
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: Try to avoid putting random newlines in when staging your changes to commit, they are just asking for conflicts later.

File: module/vision/GoalDetector/src/GoalDetector.cpp [hunk 6]
```diff
@@ -423,6 +466,7 @@ namespace vision {
                         it->measurement.push_back(Goal::Measurement(Goal::MeasurementType::TOP_NORMAL, top));
                     }
 
+
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.cpp]: newline

File: module/vision/GoalDetector/src/GoalDetector.h [hunk 7]
```diff
@@ -38,12 +39,16 @@ namespace vision {
         double MAXIMUM_GOAL_HORIZON_NORMAL_ANGLE;
         double MAXIMUM_ANGLE_BETWEEN_GOALS;
         double MAXIMUM_VERTICAL_GOAL_PERSPECTIVE_ANGLE;
+        arma::running_stat<double> stats;
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.h]: what is this doing on the header? it isn't used anywhere as far as I can tell

File: module/vision/GoalDetector/src/GoalDetector.h [hunk 8]
```diff
@@ -38,12 +39,16 @@ namespace vision {
...
 
         uint MEASUREMENT_LIMITS_LEFT;
         uint MEASUREMENT_LIMITS_RIGHT;
         uint MEASUREMENT_LIMITS_TOP;
         uint MEASUREMENT_LIMITS_BASE;
 
+        arma::vec3 vector3_covariance;
```
Comment (TrentHouliston) [inline module/vision/GoalDetector/src/GoalDetector.h]: Match the style

File: shared/utility/math/vision.h [hunk 9]
```diff
@@ -72,9 +72,16 @@ namespace math {
             return (separation / 2) / std::tan(parallaxAngle / 2);
         }
 
+        inline double distanceToEquidistantCamPoints(const double& separation,
```
Comment (TrentHouliston) [inline shared/utility/math/vision.h]: This could use some comments

File: shared/utility/math/filter/ParticleFilter.h (hunk 2 above)
Comment (TrentHouliston) [inline shared/utility/math/filter/ParticleFilter.h]: Updated @josiahw's comment to have better code.
The way that it's done at the moment join_cols will be making a new matrix every time it's called. That'll get pretty expensive over time.

File: module/localisation/RobotParticleLocalisation/src/RobotModel.h [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+        arma::vec3 getCylindricalPostCamSpaceNormal(const message::vision::Goal::MeasurementType& type,
+                                                    const arma::vec3& post_centre,
+                                                    const utility::math::matrix::Transform3D& Hcf,
+                                                    const message::support::FieldDescription& fd);
+        arma::vec3 getSquarePostCamSpaceNormal(const message::vision::Goal::MeasurementType& type,
+                                               const arma::vec3& post_centre,
+                                               const utility::math::matrix::Transform3D& Hcf,
+                                               const message::support::FieldDescription& fd);
+        // TODO: use these again?
```
Comment (TrentHouliston) [inline module/localisation/RobotParticleLocalisation/src/RobotModel.h]: Should these be deleted?

File: module/localisation/RobotParticleLocalisation/src/RobotModel.cpp [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        auto state2 = state;
+        // state2[kAngle] = normalizeAngle(state2[kAngle]);
+        // TODO: Clip robot's state to the field?
+        return state2;
+    }
+
+    arma::mat::fixed<RobotModel::size, RobotModel::size> RobotModel::processNoise() {
+        arma::mat noise = arma::eye(size, size);
+        // TODO: this
+        // noise(kX, kX) *= cfg_.processNoisePositionFactor;
```
Comment (TrentHouliston) [inline module/localisation/RobotParticleLocalisation/src/RobotModel.cpp]: This looks like it should be deleted

File: module/localisation/RobotParticleLocalisation/src/RobotModel.cpp [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+
+        // rZFf = vector from field origin to zenith high in the sky
+        arma::vec3 rZFf = {0, 0, 1};
+        arma::vec3 rZCc = Hcf.transformVector(rZFf);
+
+        if (type == Goal::MeasurementType::CENTRE) {
+            // rGCc = vector from camera to goal post expected position
+            arma::vec3 rGCc     = Hcf.transformPoint(actual_position);
+            arma::vec3 rGCc_sph = cartesianToSpherical(rGCc);  // in r,theta,phi
+            // std::cout << "actual_position = " << actual_position.t() << std::endl;
```
Comment (TrentHouliston) [inline module/localisation/RobotParticleLocalisation/src/RobotModel.cpp]: can you delete all these couts now that it's working?

File: module/localisation/BallLocalisation/src/BallModel.cpp [hunk 13]
```diff
@@ -66,11 +66,12 @@ namespace localisation {
     }
 
     arma::vec::fixed<BallModel::size> BallModel::limitState(const arma::vec::fixed<size>& state) const {
-        return state;
+        // TODO: configure
```
Comment (TrentHouliston) [inline module/localisation/BallLocalisation/src/BallModel.cpp]: This looks like something that will come back to bite us if we don't deal with it now. Nobody is going to remember this thing in a few weeks/months time.

File: module/behaviour/skills/HeadBehaviourSoccer/src/Searcher.h [hunk 14]
```diff
@@ -76,7 +76,7 @@ namespace behaviour {
                 // Full sort
                 // std::vector<std::pair<int, T>> relativePoints;
                 // for (uint i = 0; i < points.size(); i++){
-                //  relativePoints.push_back(std::pair<int, T>(i,points[i] - refPoint));
+                // 	relativePoints.push_back(std::pair<int, T>(i,points[i] - refPoint));
```
Comment (TrentHouliston) [inline module/behaviour/skills/HeadBehaviourSoccer/src/Searcher.h]: ??? why is this changed

File: module/behaviour/planning/BezierWalkPathPlanner/src/BezierWalkPathPlanner.cpp [hunk 15]
```diff
@@ -356,6 +354,7 @@ namespace behaviour {
                             // emit(graph("distanceToBall", distanceToBall));
                             // emit(graph("forwardSpeed2", finalForwardSpeed));
 
+
```
Comment (TrentHouliston) [inline module/behaviour/planning/BezierWalkPathPlanner/src/BezierWalkPathPlanner.cpp]: random whitespace is asking for future conflicts

File: Vagrantfile [hunk 16]
```diff
@@ -23,6 +23,9 @@ Vagrant.configure("2") do |config|
     # Use custom box because official Ubuntu one is shit.
     override.vm.box = "bidski/xenial64"
 
+
+    override.vm.boot_timeout = 360
```
Comment (TrentHouliston) [inline Vagrantfile]: this shouldn't be in this pr

File: shared/utility/math/filter/ParticleFilter.h (hunk 2 above)
Comment (JakeFountain) [inline shared/utility/math/filter/ParticleFilter.h]: This issue will be fixed when Walker/RobotLocalisation is pulled into master with other more significant changes

File: Vagrantfile (hunk 16 above)
Comment (JakeFountain) [inline Vagrantfile]: I thought this was put in master ages ago... @Bidski
